### PR TITLE
New module to control launchd services on target macOS hosts.

### PR DESCRIPTION
**Pull request by martinm82**
##### SUMMARY
This PR is about introducing a new core module that can be used to control launchd services on macOS targets.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
launchd

##### ANSIBLE VERSION
Tested with
```
ansible 2.4.2.0
```

### CODE REVIEW

**Comment by mattclay** [comment]
@martinm82 Can you include an integration test for this new module?

**Comment by martinm82** [comment]
@mattclay, yes will do as soon as I figured out how to create tests and run them. But will do that for sure.

BTW, I know there is already a PR open (https://github.com/ansible/ansible/pull/20881) and I actually tried to use that module in the beginning but faced couple of issues and started actually to rewrite it.

**Comment by martinm82** [comment]
@mattclay, do you have somewhere an example on how to write an integration test for macOS? I see a lot under the tests/integration folder but how do I ensure that the test will be executed only on macOS?

Shall I add just some conditions (aka ansible_family == 'Darwin') to my tasks? Or is there another way of controlling this?

Maybe you have some readme for writing integration tests. This one (https://docs.ansible.com/ansible/2.6/dev_guide/testing_integration.html) is not really helpful.

Thanks a lot

File: lib/ansible/modules/system/launchd.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+        for line in out.splitlines():
+            if line.strip():
+                pid, last_exit_code, label = line.split('\t')
+                if label.strip() == self._service:
+                    service_pid = pid
+                    status_code = last_exit_code
+                    if last_exit_code != '0':
+                        # Something strange happened and we have no clue in
+                        # which state the service is now. Therefore we mark
+                        # the service as UNKNOWN.
```
Comment (martinm82) [inline lib/ansible/modules/system/launchd.py]: ... the service state ...

**Comment by martinm82** [comment]
Seems that I somehow figured out how to write integration tests but the only thing I don't understand is why I have to set in `aliases` `shippable/posix/group[1-3]` instead of `posix/ci/group[1-3]` as it is done in all other tests.

**Comment by mattclay** [comment]
@martinm82 The `posix/ci/group[1-3]` aliases were replaced with `shippable/posix/group[1-3]`. If you haven't rebased your branch in a while then you'll still see the old aliases.

File: test/integration/targets/launchd/aliases [hunk 2]
```diff
@@ -0,0 +1 @@
+shippable/posix/group1
```
Comment (mattclay) [inline test/integration/targets/launchd/aliases:1]: Also add:

```
skip/freebsd
skip/rhel
```

Comment (martinm82) [reply test/integration/targets/launchd/aliases:1]: Fixed

**Comment by martinm82** [comment]
The current coverage of the module is at 76%:

`lib/ansible/modules/system/launchd.py                           210     36     66     14    76%`

Will try to increase it.


EDIT:
increased to 
`lib/ansible/modules/system/launchd.py | 210 | 28 | 0 | 66 | 16 | 80%`

**Comment by martinm82** [comment]
@mattclay I removed the WIP tag from the title as I think that the module is in a quite good shape including as well some integration tests.

**Comment by martinm82** [comment]
Hi guys,
it is not super urgent to get this change into master but I was just curious how the procedure with such contributions is in general.
cheers

**Comment by martinm82** [comment]
@bcoca, since I got inspired by your PR you started some time ago already maybe you could take a look at this one whether this module fits your needs as well.

**Comment by mattclay** [comment]
@martinm82 Having members of the Ansible community who are interested in this module test and review your PR and provide their feedback would be a good next step. I recommend asking for feedback in the #ansible-devel IRC channel on Freenode.

**Comment by petemounce** [comment]
@mattclay @martinm82 - we are using this module within our own ansible 2.7.5 based automation, and it appears to work for us. The tasks:

```yaml
- name: Deploy custom node_exporter job definition
  template:
    src: improbable.prometheus.node_exporter.plist.j2
    dest: "/Library/LaunchDaemons/improbable.prometheus.node_exporter.plist"
  become: yes

# Using custom module launchd https://github.com/ansible/ansible/pull/45541
- name: Make sure node_exporter is started
  launchd:
    name: improbable.prometheus.node_exporter
    state: started
  become: yes
```

The plist template:
```plist

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key>
  <string>improbable.prometheus.node_exporter</string>
  <key>ProgramArguments</key>
  <array>
    <string>{{ node_exporter_install_dir }}/node_exporter-{{ node_exporter_version }}.{{ node_exporter_platform_architecture[ansible_os_family] }}/node_exporter</string>
    <string>{% for config in node_exporter_service_config[ansible_os_family] if config[1] is not none %}--{{ config[0] }}='{{ config[1] }}' {% endfor %} {% for collector in node_exporter_collector_____enabled_collectors[ansible_os_family] %}--{{ collector }} {% endfor %}</string>
  </array>
  <key>UserName</key>
  <string>{{ node_exporter_owner }}</string>
  <key>GroupName</key>
  <string>{{ node_exporter_group }}</string>
  <key>RunAtLoad</key>
  <true/>
  <key>KeepAlive</key>
  <true/>
  <key>WorkingDirectory</key>
  <string>{{ node_exporter_install_dir }}/node_exporter-{{ node_exporter_version }}.{{ node_exporter_platform_architecture[ansible_os_family] }}</string>
  <key>StandardErrorPath</key>
  <string>/tmp/node_exporter.log</string>
  <key>StandardOutPath</key>
  <string>/tmp/node_exporter.log</string>
</dict>
</plist>
```

**Comment by Akasurde** [comment]
@dagwieers Could you please review this ? LGTM

**Comment by dagwieers** [comment]
Rather then asking for a new round of changes (especially given that this one was waiting for feedback for a longer time), I opted to make those changes myself (since they only affect the docs).

**Comment by dagwieers** [comment]
rebuild_merge

**Comment by dagwieers** [comment]
@Akasurde I think we need to move modules like this one under a separate directory 'macos' so that new modules added to this directory is automatically labeled correctly. Same for some other OS-specific modules.

**Comment by dagwieers** [comment]
@martinm82 @petemounce We have recently started a macOS Working Group to make it easier for macOS users to collaborate on macOS-specific technologies. You can join by adding yourself to the list: https://github.com/ansible/ansible/wiki/macOS

**Comment by dagwieers** [comment]
@martinm82 @petemounce We have recently started a macOS Working Group to make it easier for macOS users to collaborate on macOS-specific technologies. You can join by adding yourself to the list: https://github.com/ansible/community/wiki/macOS

**Comment by dagwieers** [comment]
There is an issue with the integration tests. I also found it quite complex in design when troubleshooting it as the actions and validation tests are not together.

The main problem in the design apparently is that `register` cannot be templated.

```
01:22  [WARNING]: "register" is not templatable, but we found: {{ action_prefix | default('test') }}_launchd_start_result, it will not be templated and will be used "as is".
```

As a result the validation test fails.

**Comment by martinm82** [comment]
Hm, this is actually a strange error as the tests were passing months ago. Was there any change in terms of what `register` accepts? I checked the documentation and could not find anything saying that register does not allow templated variables.

Regarding the complexity of the tests. Yes, I needed to make a tradeoff between readability of the tests itself and code duplication.

**Comment by martinm82** [comment]
This simple playbook works with Ansible 2.6.3:

```
- name: Test
  hosts: localhost
  connection: local
  gather_facts: no
  vars:
    - var_name: test
  tasks:
    - shell: ls -lrt
      register: "{{ var_name }}_1"
    - debug: msg="{{ test_1 }}"
```

**Comment by martinm82** [comment]
I actually did not check what your guidelines are with regards to merging with upstream. In another project I got told to do a rebase and push force which I did here as well.

@dagwieers, sorry for doing that in case it is against your rules.

**Comment by dagwieers** [comment]
@martinm82 Sometimes you need to rebase and force-push. But it doesn't help reviewing (i.e. it means we have to review everything again ;-)). No worries.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/111043/36/tests

**Comment by martinm82** [comment]
@dagwieers I have updated the integration tests and decided to remove the action tasks. In the end the duplication is not that bad.

The tests are passing now.

Would appreciate it to get this reviewed and push forward.

File: test/integration/targets/launchd/README.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Launchd integration tests
```
Comment (mattclay) [inline test/integration/targets/launchd/README.md]: Existing documentation covers how to run integration tests, so this file isn't needed.

See: https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html

File: test/sanity/code-smell/use-argspec-type-path.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
```
Comment (mattclay) [inline test/sanity/code-smell/use-argspec-type-path.py]: This file should not be part of the PR. I'm guessing it was accidentally pulled in during a rebase.

Comment (martinm82) [reply test/sanity/code-smell/use-argspec-type-path.py]: @mattclay after removing this file this test started to fail again: https://github.com/ansible/ansible/pull/45541#issuecomment-576212862

**Comment by luizribeiro** [comment]
Thanks for this PR. This is really good and I've been using it already :)

Shouldn't there also be a `loaded` `state` available? For example, I have a `local.backup.plist` [file](https://gist.github.com/luizribeiro/74aecee4088edc8500ca6ab05cd34cc9) under my `/Library/LaunchAgents` that periodically does backups. When provisioning my machine, I just want to make sure that launch agent is loaded properly, which I was hoping I could do with something like this:

```
- name: Backups are enabled
  launchd:
    name: local.backup
    state: loaded
  become: yes
```

Or is the idea to use `reloaded` in that case? I tried with `reloaded` and it seems to work fine, but I believe it will run unnecessary commands since nothing has changed. I thought I'd check for changes on the task that sets up the `.plist` file and only force a reload when that changes.

**Comment by martinm82** [comment]
@luizribeiro, happy to hear that I am not the only one using this module :)

You are right, I did not thought about this scenario as I was mainly thinking that `started` == `loaded`. 
Extending the code should be easy with this additional state. I will get back to it once all the tests and checks are passing here.

File: test/sanity/code-smell/use-argspec-type-path.py (hunk 4 above)
Comment (mattclay) [reply test/sanity/code-smell/use-argspec-type-path.py]: You will need to add an entry to the `test/sanity/ignore.txt` file:

```
lib/ansible/modules/system/launchd.py use-argspec-type-path
```

**Comment by luizribeiro** [comment]
I was getting a `FileNotFoundError` while using this with a `.plist` file that is installed under `~/Library/LaunchAgents/`:

```
Traceback (most recent call last):
  File "master:/usr/local/lib/python3.8/site-packages/ansible_mitogen/runner.py", line 975, in _run
    self._run_code(code, mod)
  File "master:/usr/local/lib/python3.8/site-packages/ansible_mitogen/runner.py", line 939, in _run_code
    exec(code, vars(mod))
  File "master:/Users/luiz/projects/ansible/library/launchd/launchd.py", line 483, in <module>
  File "master:/Users/luiz/projects/ansible/library/launchd/launchd.py", line 451, in main
  File "master:/Users/luiz/projects/ansible/library/launchd/launchd.py", line 158, in __init__
  File "master:/Users/luiz/projects/ansible/library/launchd/launchd.py", line 184, in __update
  File "master:/Users/luiz/projects/ansible/library/launchd/launchd.py", line 206, in __handle_param_force_stop
  File "/usr/local/Cellar/python@3.8/3.8.1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/plistlib.py", line 99, in readPlist
    with _maybe_open(pathOrFile, 'rb') as fp:
  File "/usr/local/Cellar/python@3.8/3.8.1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/usr/local/Cellar/python@3.8/3.8.1/Frameworks/Python.framework/Versions/3.8/lib/python3.8/plistlib.py", line 82, in _maybe_open
    with open(pathOrFile, mode) as fp:
FileNotFoundError: [Errno 2] No such file or directory: '~/Library/LaunchAgents/local.brew-update-notifier.plist'
```

It turns out we just have to make sure `Plist.__find_service_plist` returns the expanded full path to the `plist` file. This patch fixed the issue for me:

```diff
diff --git a/lib/ansible/modules/system/launchd.py b/lib/ansible/modules/system/launchd.py
--- a/library/launchd/launchd.py
+++ b/library/launchd/launchd.py
@@ -169,8 +169,9 @@ class Plist:
         ]

         for path in launchd_paths:
+            path = os.path.expanduser(path)
             try:
-                files = os.listdir(os.path.expanduser(path))
+                files = os.listdir(path)
             except OSError:
                 continue
```

**Comment by martinm82** [comment]
@luizribeiro, is it ok if I apply your patch or do you want to contribute to this PR directly?

**Comment by luizribeiro** [comment]
Feel free to apply it.

The difference is that `__find_service_plist` will return the expanded path to the `.plist` file. For example, `/Users/luizribeiro/Library/LaunchAgents/local.foo.plist' as opposed to `~/Library/LaunchAgents/local.foo.plist' which it was originally.

**Comment by martinm82** [comment]
@luizribeiro but my code does that already:

```
files = os.listdir(os.path.expanduser(path))
```

The only difference is that I am not storing it in path again.

**Comment by luizribeiro** [comment]
I can take a closer look at it over the weekend, but FWIW the tasks I was having trouble with looked like this:

```yaml
- name: Setup local.brew-update-notifier LaunchAgent
  copy:
    src: files/local.brew-update-notifier.plist
    dest: "{{ ansible_env.HOME }}/Library/LaunchAgents/local.brew-update-notifier.plist"
- name: local.brew-update-notifier Launch Agent is loaded
  launchd:
    name: local.brew-update-notifier
    state: reloaded
```

**Comment by martinm82** [comment]
ready_for_review

**Comment by martinm82** [comment]
ready_for_review

**Comment by mkrizek** [comment]
!waffling ci_verified

**Comment by jctanner** [comment]
!waffling stale_ci

**Comment by martinm82** [comment]
Ok, it's a bit frustrating to have this PR open for almost 1.5 year and seeing no progress in being able to merge it to devel branch.

Since it takes so long to contribute I will publish this module as an Ansible Galaxy module instead. This will of course mean that I won't be able to easily port the integration tests anymore.

**Comment by mattclay** [comment]
@martinm82 Now that most modules have been migrated to collections, I recommend opening a PR for this collection: https://github.com/ansible-collections/community.general/

You'll be able to use the same code, tests, etc.

Keep in mind things are still being sorted out with many of the migrated collections, so it may be a while before they're ready to merge new feature PRs.

**Comment by luizribeiro** [comment]
FWIW this module has been very helpful for my own deployments, so I would love to see this in the main distribution as opposed to Galaxy - as we have similar modules for managing services on most distributions, it seems like that would make sense since macOS is a widely supported distribution for Ansible.

**Comment by sivel** [comment]
@luizribeiro starting with ansible 2.10, only systemd, service, and sysvinit will exist in this repository. around 95% of all modules/plugins have been removed, and placed into smaller collection repos.  No new modules or plugins are being accepted in this repo.  If you have more questions feel free to swing by IRC.

**Comment by luizribeiro** [comment]
Ahhh that makes sense. The main distro was probably getting bloated. Thanks for the context!

**Comment by martinm82** [comment]
Closing this PR in favor of https://github.com/ansible-collections/community.general/pull/305
